### ISSUE TITLE
bugprone-narrowing-conversions warns even when no narrowing is for sure happening

### ISSUE THREAD

**Issue by llvmbot** [llvm/llvm-project#42931]
|  |  |
| --- | --- |
| Bugzilla Link | [43586](https://llvm.org/bz43586) |
| Version | unspecified |
| OS | Linux |
| Reporter | LLVM Bugzilla Contributor |
| CC | @legrosbuffle,@EugeneZelenko |

## Extended Description

In the following code

```c++
#include <cstdio>
#include <limits>

int main(int argc, char *argv[])
{
    unsigned int bla = std::numeric_limits<unsigned int>::max();
    int foo = bla % 10;
    printf("%d\n", foo == argc);
}
```

we get

```
main.cpp:7:15: warning: narrowing conversion from 'unsigned int' to signed type 'int' is implementation-defined [bugprone-narrowing-conversions]
    int foo = bla % 10;
```

But we know foo will be in the 0..9 range, so the conversion to int is always well defined, isn't it??

**Comment by legrosbuffle** [comment]
We don't do value range analysis in this check, the analysis is purely type-based. This is a know false positive, we don't have plans to fix that.
